### ISSUE TITLE
False positive: isinstance narrowing to np.ndarray leaves impossible int/float & ndarray variants, causing ndarray.tolist overload failure

### ISSUE THREAD

**Issue by myrlagksruf** [astral-sh/ty#3707]
### Summary

`ty` appears to keep impossible intersection variants after an `isinstance(..., np.ndarray)` check. This causes a valid `ndarray.tolist()` call to fail with `no-matching-overload`.

### Minimal repro

```py
import numpy as np


def f(sliced_values: int | float | np.ndarray) -> None:
    assert isinstance(sliced_values, np.ndarray)
    assert sliced_values.tolist()[3:5] == [0.13, 0.14]
```

### Expected behavior

After the assertion, `sliced_values` should be narrowed to `np.ndarray`, so calling `sliced_values.tolist()` with no arguments should be accepted.

### Actual behavior

`ty` reports `no-matching-overload` for `ndarray.tolist`. The diagnostic suggests that `ty` is still considering impossible variants such as `int & ndarray[...]` and `float & ndarray[...]` after the narrowing:

```text
No overload of bound method `ndarray.tolist` matches arguments

info: Possible overloads for bound method `tolist`:
info:   (self: ndarray[Never, Unknown], /) -> Any
info:   [_T](self: ndarray[tuple[()], Unknown], /) -> _T
info:   [_T](self: ndarray[tuple[int], Unknown], /) -> list[_T]
info:   [_T](self: ndarray[tuple[int, int], Unknown], /) -> list[list[_T]]
info:   [_T](self: ndarray[tuple[int, int, int], Unknown], /) -> list[list[list[_T]]]
info:   (self, /) -> Any
info: Union variant `bound method int & ndarray[tuple[object, ...], dtype[object]].tolist() -> Any` is incompatible with this call site
info: Attempted to call union type `(bound method int & ndarray[tuple[object, ...], dtype[object]].tolist() -> Any) | (bound method float & ndarray[tuple[object, ...], dtype[object]].tolist() -> Any) | (Overload[[_T]() -> _T, [_T]() -> list[_T], [_T]() -> list[list[_T]], [_T]() -> list[list[list[_T]]], () -> Any])`
```

### Notes

This looks like a false positive in narrowing / impossible-intersection pruning rather than a NumPy runtime issue. At runtime, the assertion guarantees that `sliced_values` is an `np.ndarray`, and `ndarray.tolist()` is a valid zero-argument method call.

A workaround is to cast after the assertion:

```py
from typing import Any, cast
import numpy.typing as npt

assert isinstance(sliced_values, np.ndarray)
arr = cast(npt.NDArray[Any], sliced_values)
assert arr.tolist()[3:5] == [0.13, 0.14]
```

### Version

0.0.45, 0.0.46

**Comment by carljm** [comment]
The [type definition for ndarray](https://github.com/numpy/numpy/blob/f149b48850d8ca4c0bc18d542d3ea573e54d0ab6/numpy/__init__.pyi#L2078) doesn't tell us that it can't be multiply-inherited with `int`. That runtime fact (due to the implementation of `ndarray` in C, and currently invisible to ty) is the only reason that `int & ndarray[...]` is impossible. This fact could be made visible to ty if ndarray were decorated with `typing_extensions.disjoint_base`, introduced by PEP 800. This would be an accurate and otherwise harmless change to the numpy stubs.

ty could potentially assume such multiple inheritance does not occur, even without knowing that it's impossible; that's #1578. But barring a more general behavior change in ty there, the adoption of `disjoint_base` in the numpy stubs is the most straightforward fix here.
